Added local store availability for variable products.

danielistrate27 opened the pull request:
### Ticket
- [Update variant availability on product page](https://app.asana.com/0/1202867909936929/1202493915593240)

danielistrate27 committed:
Changed variant availability table content.

danielistrate27 committed:
Fixed default view of the table.

danielistrate27 committed:
Updated hide_variation event and moved styles from theme to plugin.

danielistrate27 committed:
Changed asterisk to be only on showroom column.

danielistrate27 committed:
Updated text under availability table.

creative-andrew commented on src/Product.php:
```diff
@@ -9,6 +9,8 @@ class Product {
 
   const CATEGORY_EXCLUDED = ['Ausstellungsstücke', 'Abverkauf'];
 
+  private static $allVariationsStock;
+
```
```suggestion
  private static $allVariationsStock;
```

Can you add documentation for this member variable?

danielistrate27 replied:
Updated in the last commit.

creative-andrew commented on src/Product.php:
```diff
@@ -214,17 +219,43 @@ public static function display_store_stock_status_block() {
+          }
+        }
+      }
+    }
```
```suggestion
    if (!empty(static::$allVariationsStock) && count(static::$allVariationsStock) > 1) {
      foreach (reset(static::$allVariationsStock) as $location => $stocks) {
        $availability[$location] = [];
        foreach ($stocks as $type => $stock) {
          if (!isset($availability[$location][$type])) {
            $availability[$location][$type] = [];
          }
        }
      }
      foreach (static::$allVariationsStock as $product_id => $locations) {
        foreach ($locations as $location => $stocks) {
          foreach ($stocks as $type => $stock) {
            if ($stock != 'none' && $type == 'showroom') {
              $availability[$location][$type][] = $product_id;
              $store_stocks[$location][$type] = 'high-asterisk';
            }
          }
        }
      }
    }
```

Is there any way to refactor this in a different method(s) including documentation? 3 levels of foreach is really hard to read.

danielistrate27 replied:
Updated in the last commit.

danielistrate27 committed:
Refactored nested foreach and added missing documentaion.

danielistrate27 committed:
Merged branch with master repo.

sun commented on languages/woocommerce-local-store.pot:
```diff
@@ -42,14 +42,18 @@ msgstr ""
 msgstr ""
+
+#: templates/store-stock.php:42
+msgid "*This product is exhibited in a different, similar variant."
```
The asterisk at the beginning of the notice "This product is exhibited in a different, similar variant." should either be separated with a space or maybe not be part of the output entirely (CSS :before instead?).

danielistrate27 replied:
Updated in the last commit.

sun commented on src/Product.php:
```diff
@@ -191,7 +198,10 @@ public static function getStockByStore(\WC_Product $product): array {
       $stocks[$name]['online'] = ($is_backorder_allowed || $warehouse_stock) ? 'high' : 'none';
     }
-
+    if (!isset(static::$allVariationsStock)) {
+      static::$allVariationsStock = [];
+    }
```
Please initialize the class property as an empty array instead.

danielistrate27 replied:
Updated in the last commit.

sun commented on src/Product.php:
```diff
@@ -9,6 +9,13 @@ class Product {
   const CATEGORY_EXCLUDED = ['Ausstellungsstücke', 'Abverkauf'];
 
+  /**
+   * Stores the stock of the products across all variations available.
```
Please add a PHPDoc description that explains for what reason this is done.

danielistrate27 replied:
Updated in the last commit.

sun commented on src/Product.php:
```diff
@@ -214,20 +224,95 @@ public static function display_store_stock_status_block() {
-    $stocks = static::getStockByStore($product);
-    $raw = $stocks;
-    foreach ($stocks as $name => $types) {
+    $availability = self::buildInitialAvailabilityArray(static::$allVariationsStock);
+    if (!empty(static::$allVariationsStock) && count(static::$allVariationsStock) > 1) {
+      foreach (static::$allVariationsStock as $product_id => $locations) {
+        $availability = self::getAvailabilityWithProductID($availability, $product_id, $locations);
+        $store_stocks = self::getStoreStocksWithUpdatedAvailability($store_stocks, $product_id, $locations);
+      }
+    }
```
This appears to prepare additional stock availability data. This should not be done within the display/render function. There are separate functions that are retrieving/generating the data only.

Note that the hook implementation `woocommerce_available_variation` right in this file already prepares the stock by store for each loaded variant. The information is output within the JSON data for all variants in the frontend. Maybe it would be faster and better to process and calculate that existing information in the frontend only?  Otherwise we really need to ensure that this change in the backend does not cause major performance bottleneck (due to loading all variants again).

danielistrate27 replied:
I've changed where the data is calculated. Moved everything in the woocommerce_available_variation filter.
Let me know if this satisfies the requested change.

danielistrate27 committed:
Loading availability array outside render function.

sun commented on src/Product.php:
```diff
@@ -202,9 +213,25 @@ public static function getStockByStore(\WC_Product $product): array {
    */
   public static function woocommerce_available_variation(array $variation_data, \WC_Product $parent, \WC_Product $variation): array {
     $variation_data['stock_levels'] = static::getStockByStore($variation);
+    $variation_data['availability'] = static::getProductAvailability();
```
1. To my knowledge, WooCommerce core already sets the key `availability` with the basic stock/availability information - I assume you did not intend to overwrite it?
2. Why aren't you passing any of the arguments as parameters to the function?  And do you really need to build the initial array?  It seems like you can skip that and call getAvailabilityWithProductID() to only build the info relevant for the product and its variants that are currently being prepared in `woocommerce_available_variation`.

sun commented on src/Stock.php:
```diff
@@ -29,6 +29,9 @@ public static function renderLevel(string $stock_level, string $type): string {
     }
+    elseif ($stock_level === 'high-asterisk') {
+      $text = __('Low stock', Plugin::L10N);
+    }
```
Why would a high stock level output a "low stock" label in the UI?  Mixing the stock level with the variation notice doesn't strike me as a good idea.  Isn't the variation notice always calculated in the JS based on the additional meta data in the available variations?  Why do we need to adjust the backend levels then?

sun commented on src/Product.php:
```diff
@@ -214,20 +241,66 @@ public static function display_store_stock_status_block() {
     global $product;
+    $store_stocks = static::getStockByStore($product);
 
-    $stocks = static::getStockByStore($product);
```
Seems like the renaming of this variable can be reverted to limit/reduce the amount of changes in this PR.

sun commented on src/Product.php:
```diff
@@ -214,20 +241,66 @@ public static function display_store_stock_status_block() {
+  public static function getAvailabilityWithProductID(array &$availability, int $product_id, array $locations) {
+    foreach ($locations as $location => $stocks) {
+      foreach ($stocks as $type => $stock) {
+        if ($stock != 'none' && $type == 'showroom') {
```
Why is only stock in a showroom considered as available?

sun commented on src/Product.php:
```diff
@@ -214,20 +241,66 @@ public static function display_store_stock_status_block() {
+   * @param array $locations
+   *   The locations array for the product_id.
+   */
+  public static function getAvailabilityWithProductID(array &$availability, int $product_id, array $locations) {
```
`$availability` should not be passed by reference.
```suggestion
  public static function getAvailabilityWithProductID(array $availability, int $product_id, array $locations) {
```

danielistrate27 committed:
Refactored code, moved the functionality in JS.

danielistrate27 commented:
@sun 
I have updated the entire functionality, and moved all the process in the JS on hide_variation and show_variation events. 
"Why is only stock in a showroom considered as available?" - This is the requested functionality. If a variation doesn't have stock in the showroom but there are other product variations that can be found in the showroom, we display a "*" with a green dot.(https://app.asana.com/0/0/1202921777945027/1203003296765399/f)

creative-andrew commented:
Not longer needed as this functionality has been moved to WOPA-FRONT
